src/controller/stream.rs

```diff
@@ -0,0 +1,341 @@
+        let first_chunk = network.find_chunk(&request.dataset_id, request.query.first_block());
+        let Some(first_chunk) = first_chunk else {
+            tracing::info!(
+                "No chunk found for block {}, dataset {}",
+                request.query.first_block(),
+                request.dataset_id
+            );
+            return Err(RequestError::NotFound(format!(
+                "No chunk found for block {}",
+                request.query.first_block()
+            )));
+        };
+
+        let buffer = SlidingArray::with_capacity(request.buffer_size);
+        let timeout_quantile = request.timeout_quantile;
+        let mut streamer = Self {
+            request,
+            stream_id,
+            network,
+            next_index: 0,
+            next_chunk: Some(first_chunk),
+            downloads: Default::default(),
+            buffer,
+            timeouts: Arc::new(TimeoutManager::new(timeout_quantile)),
+            trial: true,
+        };
+
+        // Intentionally schedule only one download not to waste resources in case of bad request
+        streamer.schedule_next_chunk();
+
+        Ok(streamer)
+    }
+
+    #[instrument(skip_all, fields(stream_id = self.stream_id))]
+    pub async fn poll_next_chunk(&mut self) -> Option<Result<ResponseChunk, RequestError>> {
+        while self.buffer.len() < self.request.buffer_size
+            && self.next_chunk.is_some()
+            && !self.trial
+        {
+            self.schedule_next_chunk();
+        }
+
+        while !self.buffer.front().is_some_and(Option::is_some) {
+            let (result, index) = self
+                .downloads
+                .join_next()
+                .await?
+                .expect("Awaiting download panicked");
+            if self.buffer.get(index).is_some_and(Option::is_none) {
+                let code = short_code(&result);
+                self.buffer[index] = Some(result);
+                let mask = self
+                    .buffer
+                    .data()
+                    .iter()
+                    .map(|slot| if slot.is_some() { '#' } else { '.' })
+                    .collect::<String>();
+                tracing::debug!(
+                    "Got response for chunk number {index} ({code}), buffer:\n[{mask}]",
+                );
+            } else {
+                tracing::debug!("Dropping duplicate response for chunk number {index}");
+            }
+        }
+
+        let result = self.buffer.pop_front()?.unwrap();
+        self.trial = false;
+        if let Ok(bytes) = &result {
+            tracing::debug!("Sending {} response bytes", bytes.len());
+        }
+        Some(result)
+    }
+
+    // Schedules downloading `self.next_chunk`
+    fn schedule_next_chunk(&mut self) {
+        let chunk = self.next_chunk.clone().expect("No next chunk to download");
+        let index: usize = self.next_index;
+        let future = self
+            .get_entire_chunk(index, chunk.clone())
```

Comment:
I can propose the following solution.

When at least one chunk for the current stream doesn't fit in the response, immediately switch to the "slow" streaming mode. In this mode only a single chunk is requested simultaneously (no fetch ahead), but the partial results are sent to the client as soon as they are received. This would probably be enough for such a use case (which is not a recommended usage anyway), but would limit the memory usage to a single response size in this mode.

Reply:
Okay, I've fixed it in a proper way. Each "slot" of the buffer now saves only one response. If the current first slot of the buffer is only a partial response, we send it to the client immediately and request the continuation.